README.rst
Urban Rivals Collection Management API (urcollectionmanager)
============================================================

A module to support reading player purchase history data and
storing it in a database.

Puchases History Use:
    #) Import `api` from `urcollectionmanager`
    #) Create a request.Session and pass it, along with
       your credentials to `session_connect_to_ur`
    #) Pass that same session (now authenticated) to
       `get_purchase_history` along with the number of
       pages to scrape.
    #) Pass the result to `convert_purchase_history`

        Optionally: Pass result to your own HTML parser.

    #) You now have a list of Purchase objects that contain
       all relevant information available from the purchase
       history page.

Missions Use:
    #) Import `api` from `urcollectionmanager`
    #) Create a request.Session and pass it, along with
       your credentials to `session_connect_to_ur`
    #) Pass that same session (now authenticated) to
       `get_missions_list` along with the mission category
       to find. An empty string returns everything.
    #) Pass the result to `convert_missions`

        Optionally: Pass result to your own HTML parser.

    #) You now have a list of Mission objects that contain
       all relevent information available from the missions
       page.

Database Use:
    #) Run `connect_and_initialize_database` if you have a
       particular database location you want to connect to.
       By default this will connect to `data/collection.sqlite`
    #) Pass a list of Purchase objects to `write_history_to_database`

        If using `convert_purchase_history`, you will need to flatten
        the list of lists

    #) Once the database has Purchase objects in it, you can
       run `get_history_from_database` to retrieve them.

    #) Similarly, you can run `write_missions_to_database` to write
       Mission objects and `get_missions_from_database` to read mission
       Mission objects from the same database as Purchases.

Dev
---
How To Use Tools:
    Poetry_
        - To run the project (preferably configure venv first)

        >>> poetry install

        - To configure your venv

        >>> python -m venv /path/to/new/virtual/environment
        >>> poetry env info

            - If `env info` does not match your path, try manually running \
              `Scripts/activate` inside that environment

        - To add a dependency (add a -D if it's only for development)

        >>> poetry add <dependency>

    Commitizen_
        >>> git add
        >>> cz c

    - More options can be found under options under Commitizen_

    PreCommit_
        Update .pre-commit-config.yaml, then run

    >>> pre-commit install
    >>> pre-commit run --all-files

    - More hook plugins can be found at PreCommitHooks_

TODO:

- Use UR-API_ to collect player data. UR-API-Example_
- Default to file output (excel or some form of that)
- Testing (pytest, Coverage-py_)
- Documentation (Sphinx_)

Not Required at this Time:

- Black_
- iSort_

.. _UR-API: https://www.urban-rivals.com/api/developer/
.. _UR-API-Example: https://github.com/Buscatrufas/UrbanRivals/blob/master/index.php
.. _Coverage-py: https://coverage.readthedocs.io/en/latest/config.html
.. _Sphinx: https://www.sphinx-doc.org/en/master/
.. _AutoPEP8: https://github.com/hhatto/autopep8#usage
.. _Black: https://github.com/psf/black#version-control-integration
.. _iSort: https://github.com/pre-commit/mirrors-isort
.. _Commitizen: https://woile.github.io/commitizen/
.. _PreCommit: https://pre-commit.com/
.. _PreCommitHooks: https://pre-commit.com/hooks.html
.. _Poetry: https://python-poetry.org/docs/cli/
.. _PypiToken: https://pypi.org/help/#apitoken

CHANGELOG.md
## 'v1.1.3 (2020-07-19')

### Fix

- **find-missions-function**: allow missions to be created when given a list of only 1 element

## 'v1.1.2 (2020-07-19')

### Refactor

- **pull-master**: update local branch

### Fix

- **mission-category**: add black market category to list of supported categories

## 'v1.1.1 (2020-06-13')

### Fix

- **mission**: fix the issue where a completed mission caused tuple unpacking to fail

## 'v1.1.0 (2020-06-07')

### Feat

- **tests**: add unit tests for missions module
- **missions**: add support for mission management

### Fix

- **lint**: fix flake8 lint failure

## 'v1.0.1 (2020-03-18')

### Fix

- **purchase**: fix purchase date incompatibility bug

## 'v1.0 (2020-03-16')

### Refactor

- **pr**: remove test databases

### Feat

- **db**: add support for reading and writing purchase history with sqlite databases
- **api**: allow screenscraping with authenticated session

### Fix

- **db**: add support for databases, particularly sqlite


urcollectionmanager/missions.py
from .mission import Mission

import itertools
from urllib.parse import urlencode
from enum import Enum
from bs4 import BeautifulSoup

BASE_MISSION_URL = "https://www.urban-rivals.com/missions/missions_page.php?"


def get_mission_url(page_num: int, category: str, mission_status: str, url) -> str:
    """Returns a url for the requested parameters"""
    url = BASE_MISSION_URL if url is None else url
    category = "all" if category == "" else category
    return url + urlencode({'page': 0 if page_num is None else page_num,
                            'category': _convert_mission_category(category).value,
                            'state': _convert_mission_status(mission_status),
                            'sortby': 'date'})


def find_missions(page: BeautifulSoup):
    """Returns list of Mission based on each row in the
    given BeautifulSoup missions page"""
    ul_list = page.find_all("ul")
    if len(ul_list) > 1:
        return [create_mission(row) for row in ul_list[2].find_all("li")]
    return [create_mission(row) for row in ul_list[0].find_all("li")]


def create_mission(row):
    """Creates Mission object based on BeautifulSoup list item object"""
    name = next(itertools.islice(row.stripped_strings, 0, 1))
    description = next(itertools.islice(row.stripped_strings, 1, 2))
    total_progress = next(itertools.islice(row.stripped_strings, 2, 3))
    try:
        progress, goal = str(total_progress).split("/")
    except BaseException:
        progress = 0
        goal = 0
    return Mission(name=name,
                   description=description,
                   progress=progress,
                   goal=goal
                   )


def _convert_mission_category(category: str) -> int:
    try:
        return MissionCategory[category.upper().replace(" ", "_")]
    except BaseException:
        # No matching category
        raise ValueError(f"{category} is not a supported mission category.")


def _convert_mission_status(mission_status: str) -> str:
    if "progress" in mission_status.lower():
        return "inprogress"
    elif "complete" in mission_status.lower() or \
         "done" in mission_status.lower():
        return "done"
    return 'all'


class MissionCategory(Enum):
    ALL = 0
    FIGHTS = 3
    EVOLUTION = 6
    COLLECTION = 7
    ALL_STARS = 8
    BANGERS = 9
    FANG_PI_CLANG = 10
    FREAKS = 11
    GHEIST = 12
    JUNGO = 13
    JUNKZ = 14
    LA_JUNTA = 15
    MONTANA = 16
    NIGHTMARE = 17
    PIRANAS = 18
    PUSSYCATS = 19
    RESCUE = 20
    ROOTS = 21
    SAKROHM = 22
    SENTINEL = 23
    SKEELZ = 24
    ULU_WATU = 25
    UPPERS = 26
    VORTEX = 27
    LEGENDARIES = 28
    DAMAGE = 29
    PILLZ = 30
    INTOXICATION = 31
    HEALING = 32
    TOURNAMENTS = 33
    SURVIVOR = 34
    ELO = 35
    BERZERK = 37
    FROZN = 39
    HURACAN = 41
    RIOTS = 42
    FLASH = 43
    LEADER = 44
    RAPTORS = 45
    HIVE = 47
    GHOSTOWN = 49
    TUTORIAL = 50
    DOMINION = 51
    KOMBOKA = 53
    PARADOX = 54
    BLACK_MARKET = 57
